### PR TITLE
Remove name delegates from MembershipApplication

### PR DESCRIPTION
**Pull request by RobertCram**
PT Story: https://www.pivotaltracker.com/story/show/148546123

**Changes proposed in this pull request:**

1) Rewrite Rspec and Cucumber tests to use the `first_name` and `last_name` attributes from the User model instead of from the MembershipApplication model.

2) Remove the `first_name` and `last_name` delegates from the MembershipApplication model

Ready for review:
@patmbolger @weedySeaDragon

### CODE REVIEW

File: features/admin-resets-users-password.feature [hunk 1]
```diff
@@ -6,19 +6,18 @@ Feature: As an admin
...
 
     Given the following users exists
-      | email               | admin |
-      | emma@happymutts.com |       |
-      | bob@snarkybarky.se  |       |
-      | admin@shf.se        | true  |
+      | first_name | email               | admin |
+      | Emma       | emma@happymutts.com |       |
+      | Bob        | bob@snarkybarky.se  |       |
+      | admin      | admin@shf.se        | true  |
```
Comment (patmbolger) [inline features/admin-resets-users-password.feature]: I don't think `first_name` is doing anything for us in these tests.  Recommend dropping this column.

File: features/admin_creates_company.feature [hunk 2]
```diff
@@ -13,10 +13,10 @@ Feature: As an admin
...
   Background:
     Given the following users exists
-      | email                      | admin |
-      | applicant_1@happymutts.com |       |
-      | applicant_3@happymutts.com |       |
-      | admin@shf.se               | true  |
+      | first_name | email                      | admin |
+      | Emma       | applicant_1@happymutts.com |       |
+      | Anna       | applicant_3@happymutts.com |       |
+      | admin      | admin@shf.se               | true  |
```
Comment (patmbolger) [inline features/admin_creates_company.feature:19]: Same comment as above.

File: features/company_shows_all_employee_categories.feature [hunk 3]
```diff
@@ -15,12 +15,12 @@ Feature: As a visitor
...
-      | lars@happymutts.com   |       |
-      | anna@happymutts.com   |       |
-      | bowser@snarkybarky.se |       |
-      | admin@shf.se          | true  |
+      | first_name | email                 | admin |
+      | Emma       | emma@happymutts.com   |       |
+      | Lars       | lars@happymutts.com   |       |
+      | Anna       | anna@happymutts.com   |       |
+      | Bowser     | bowser@snarkybarky.se |       |
+      | admin      | admin@shf.se          | true  |
```
Comment (patmbolger) [inline features/company_shows_all_employee_categories.feature:23]: Same comment as above.

File: features/delete-company.feature [hunk 4]
```diff
@@ -14,21 +14,21 @@ Feature: As an admin
...
-      | lars@snarkybarky.com           |       |
-      | bob@bowsers.com                |       |
-      | kitty@kitties.com              |       |
-      | meow@kitties.com               |       |
-      | under_review@kats.com          |       |
-      | ready_for_review@kats.com      |       |
-      | waiting_for_review@kats.com    |       |
-      | new@kats.com                   |       |
-      | waiting_for_applicant@kats.com |       |
-      | admin@shf.se                   | true  |
```
Comment (patmbolger) [inline features/delete-company.feature:31]: Same comment ....

File: features/disallow-company-malicious-website.feature [hunk 5]
```diff
@@ -7,9 +7,9 @@ Feature: Don't allow malicious code in Company website value
 
   Background:
     Given the following users exists
-      | email               | admin |
-      | emma@happymutts.com |       |
-      | admin@shf.se        | true  |
+      | first_name | email               | admin |
+      | Emma       | emma@happymutts.com |       |
+      | admin      | admin@shf.se        | true  |
```
Comment (patmbolger) [inline features/disallow-company-malicious-website.feature:12]: You guessed it .... :)

File: features/edit_company.feature [hunk 6]
```diff
@@ -4,20 +4,20 @@ Feature: As a member
...
   Background:
     Given the following users exists
-      | email                      | admin | is_member |
-      | applicant_1@happymutts.com |       | true      |
-      | applicant_3@happymutts.com |       | false     |
-      | admin@shf.se               | true  | true      |
+      | first_name | email                      | admin | is_member |
+      | Emma       | applicant_1@happymutts.com |       | true      |
+      | Anna       | applicant_3@happymutts.com |       | false     |
+      | admin      | admin@shf.se               | true  | true      |
```
Comment (patmbolger) [inline features/edit_company.feature:10]: Ditto ...

File: features/export-member-email-info.feature [hunk 7]
```diff
@@ -6,10 +6,10 @@ Feature: export member information to a CSV file so I can use it in other system
...
   Background:
     Given the following users exists
-      | email                       | admin |
-      | emma@happymutts.com         |       |
-      | wils@woof.com               |       |
-      | admin@shf.se                | true  |
+      | first_name | last_name   | email                       | admin |
+      | Emma       | Emmasdottir | emma@happymutts.com         |       |
+      | Wils       | Wilson      | wils@woof.com               |       |
+      | admin      | admin       | admin@shf.se                | true  |
```
Comment (patmbolger) [inline features/export-member-email-info.feature:12]: These two columns are not needed either.

File: features/log_in.feature [hunk 8]
```diff
@@ -4,16 +4,16 @@ Feature: As a registered user
...
-      | email                | password | admin | is_member |
-      | emma@random.com      | password | false | true      |
-      | anne@random.com      | password | false | false     |
-      | lars-user@random.com | password | false | false     |
-      | arne@random.com      | password | true  | true      |
+      | first_name | email                | password | admin | is_member |
+      | Emma       | emma@random.com      | password | false | true      |
+      | Lars       | lars-user@random.com | password | false | false     |
+      | Anne       | anne@random.com      | password | false | false     |
+      | Arne       | arne@random.com      | password | true  | true      |
```
Comment (patmbolger) [inline features/log_in.feature:11]: Same comment.

File: features/manage_users.feature [hunk 9]
```diff
@@ -4,11 +4,11 @@ Feature: As an admin
...
-      | email               | admin |
-      | emma@happymutts.com |       |
-      | anna@sadmutts.com   |       |
-      | ernt@mutts.com      |       |
-      | admin@shf.se        | true  |
+      | first_name | email               | admin |
+      | Emma       | emma@happymutts.com |       |
+      | Anna       | anna@sadmutts.com   |       |
+      | Ernt       | ernt@mutts.com      |       |
+      | admin      | admin@shf.se        | true  |
```
Comment (patmbolger) [inline features/manage_users.feature:11]: Ditto.

File: features/mapping_and_geocoding/companies_are_geocoded_when_viewing_all.feature [hunk 10]
```diff
@@ -23,20 +23,20 @@ Feature: All companies are geocoded before being shown on the view all companies
...
 
     And the following users exists
-      | email               | admin |
-      | emma@happymutts.com |       |
-      | a@happymutts.com    |       |
-      | admin@shf.se        | true  |
+      | first_name | email               | admin |
+      | Emma       | emma@happymutts.com |       |
+      | Anna       | a@happymutts.com    |       |
+      | admin      | admin@shf.se        | true  |
```
Comment (patmbolger) [inline features/mapping_and_geocoding/companies_are_geocoded_when_viewing_all.feature:29]: Ditto.

File: features/member_edits_company_page.feature [hunk 11]
```diff
@@ -6,9 +6,9 @@ Feature: As a member
 
   Background:
     Given the following users exists
-      | email               | admin | is_member |
-      | emma@happymutts.com |       | true      |
-      | admin@shf.se        | true  | true      |
+      | first_name | email               | admin | is_member |
+      | Emma       | emma@happymutts.com |       | true      |
+      | admin      | admin@shf.se        | true  | true      |
```
Comment (patmbolger) [inline features/member_edits_company_page.feature:11]: Ditto.

File: spec/models/membership_application_spec.rb [hunk 12]
```diff
@@ -107,14 +107,14 @@
...
       }.to raise_exception(/#{I18n.t('activerecord.attributes.membership_application.first_name')} #{I18n.t('errors.messages.blank')}/)
     end
 
     it 'validates the presence of last_name' do
       expect {
-        member_app.last_name = ''
+        user.last_name = ''
         member_app.save!
       }.to raise_exception(/#{I18n.t('activerecord.attributes.membership_application.last_name')} #{I18n.t('errors.messages.blank')}/)
     end
```
Comment (patmbolger) [inline spec/models/membership_application_spec.rb:120]: These two tests should be moved to the User model spec file.

File: features/delete-company.feature [hunk 13]
```diff
@@ -14,21 +14,21 @@ Feature: As an admin
...
+      | Lars             | lars@snarkybarky.com           |       |
+      | Kitty            | kitty@kitties.com              |       |
+      | Meow             | meow@kitties.com               |       |
+      | Under_Review     | under_review@kats.com          |       |
+      | Ready for Review | ready_for_review@kats.com      |       |
+      | Waiting for A    | waiting_for_applicant@kats.com |       |
+      | New              | new@kats.com                   |       |
+      | Waiting for R    | waiting_for_review@kats.com    |       |
+      | Bob              | bob@bowsers.com                |       |
+      | admin            | admin@shf.se                   | true  |
```
Comment (patmbolger) [inline features/delete-company.feature:31]: Same comment.

File: features/membership-application-status/admin-manage-reasons-waiting-for-info.feature [hunk 14]
```diff
@@ -12,10 +12,10 @@ Feature: Admin manages the list of reasons why SHF is waiting for info from an a
...
   Background:
     Given the following users exists
-      | email                                  | admin |
-      | anna_waiting_for_info@nosnarkybarky.se |       |
-      | emma@happymutts.se                     |       |
-      | admin@shf.se                           | true  |
+      | first_name   | email                                  | admin |
+      | AnnaWaiting  | anna_waiting_for_info@nosnarkybarky.se |       |
+      | EmmaAccepted | emma@happymutts.se                     |       |
+      | admin        | admin@shf.se                           | true  |
```
Comment (patmbolger) [inline features/membership-application-status/admin-manage-reasons-waiting-for-info.feature:18]: Same comment.

**Review by patmbolger — COMMENTED** [review]
Please see comments.

File: spec/models/membership_application_spec.rb (hunk 12 above)
Comment (RobertCram) [reply spec/models/membership_application_spec.rb:120]: Not yet in this PR. At the moment we still need to make sure you cannot save a membership application without a first name and lastname.

**Comment by patmbolger** [comment]
Unfortunately there seems to have been a practice to use `first_name` of a user as a look-up key for many tests.  This has largely gone away (which is good), but a test or two probably still uses this.

Another reason to specify `first_name` in a test background is if the test looks for that name in the browser window.  One or two tests probably do this.

For the rest, moving `first_name` from the membership_application background table to the user background table just moves useless data that clutters the test.

I made remarks on many of the test files that reflect this - I gave up after a point so other tests should be checked as well.

In any case, it's your call if you want to remove the `first_name` column from feature files where it is not needed.  I do think it would be good to slim the test data down to the minimum required for each test.

File: spec/models/membership_application_spec.rb (hunk 12 above)
Comment (patmbolger) [reply spec/models/membership_application_spec.rb:120]: Ah yes, you're right.  Another "chicken and egg" scenario.

**Comment by RobertCram** [comment]
I understand Patrick. I wanted to change the tests as little as needed, but also noticed they could use a clean up. In hindsight it probably would have been better to use this PR for that. But as it stands now, I'd rather make it a new story. Is that ok with you?

**Comment by patmbolger** [comment]
@RobertCram - absolutely, a new story makes sense since this one is ready to go.

**Review by patmbolger — APPROVED** [review]

**Review by weedySeaDragon — APPROVED** [review]

**Pull request merged**